Enhance information for REST Hook triggers (#312)
zapier/visual-builder: docs/_docs/triggers.md

Polling API results are expected to be an array with 0 or more objects that will be passed to Zapier's [deduper](https://platform.zapier.com/docs/dedupe). The deduper will return any items that haven't been received before, and use them to run the subsequent steps in the user's Zap.

### Polling Trigger

Zapier selects a _Polling_ trigger by default, where Zapier will send a request to an API endpoint URL to request new data—and in live Zaps, Zapier automatically deduplicates and sorts for the newest data from the API. Live Zaps automatically poll the URL for new data every 1 to 15 minutes, depending on the user's Zapier plan (see [Plans & Pricing](https://zapier.com/pricing) for more details).

To add a polling trigger, select _Polling_ at the top of the settings page, then enter your API URL in the _API Endpoint_ field. If your API URL requires specific data in the URL path, enter the following text in the URL where your API expects that data, replacing `key` with the input field key containing the relevant data from the input form you created in the previous step:

{% raw %}`{{bundle.inputData.key}}`{% endraw %}

> Otherwise, Zapier will automatically include any input field data with the API call as URL parameters (for GET requests) or in the request body (for POST requests, e.g. for GraphQL APIs).

Reviewer: I don't think we do GraphQL by default. Suggested change:

```suggestion
Otherwise, Zapier will automatically encode any input field data with the API call as URL parameters (for GET requests) or in the request body with JSON (for POST requests).
```

Author: Yeah, that's a good point. I used that as an example because GraphQL is the most familiar type of API for me that might use POSTs for polling, but it does give the wrong impression about the level of support.

I'm curious about why you mention `encode` - is that just because we'll also default to URL-encoding the params? From the point of view of the developer, we're including them in the request and the encoding is transparent, so I'm not sure about the best word to use.

![](https://zappy.zapier.com/CF1A11AF-949A-4C74-AFCD-37F4F4C5B362.png)

We recommend that a successful Subscribe response return a 201. The data returned should include any data needed to attempt a later Unsubscribe request. This data will be stored in `bundle.subscribeData`.

#### Unsubscribe

When Zapier sends the request to your API to unsubscribe the webhook, it can reference any data that was returned during the Subscribe request and stored in `bundle.subscribeData`.

![](https://zappy.zapier.com/EF2923B4-B361-45A6-983C-BA2A57DC5623.png)

#### Perform List

In addition to the Subscribe and Unsubscribe requests, it's important to add a Perform List request where Zapier can check for recent items. This will be used to fetch data when users are setting up and testing the Zap. If you don't define a Perform List, then the user needs to go into your app and do something to generate a new event while the Zap editor waits for data, which is not an optimal experience.

<a id="perform"></a>
#### Perform

![REST Hook Perform](https://cdn.zapier.com/storage/photos/5c6c154be13fbcad2a5dc8cb786f2dea.png)

> Finally, in the Perform, you can customize the code to evaluate the data your app's webhooks pass to Zapier. By default, Zapier includes `return [bundle.cleanedRequest];` to return the request data from the webhook as an array. If your data needs to be transformed, or includes multiple objects, add custom code to parse the response data in `bundle.cleanedRequest` within the Perform and turn it into an array of objects, such as [this example code](https://github.com/zapier/zapier-platform/blob/master/example-apps/rest-hooks/triggers/recipe.js#L42).

Reviewer: Let's pin the link to a specific commit in case the file is moved one day:

```suggestion
Finally, in the Perform, you can customize the code to evaluate the data your app's webhooks pass to Zapier. By default, Zapier includes `return [bundle.cleanedRequest];` to return the request data from the webhook as an array. If your data needs to be transformed, or includes multiple objects, add custom code to parse the response data in `bundle.cleanedRequest` within the Perform and turn it into an array of objects, such as [this example code](https://github.com/zapier/zapier-platform/blob/512f558ffa6dff11a0985c2e43c159d534bb6f36/example-apps/rest-hooks/triggers/recipe.js#L42).
```

> Both the Perform and Perform List methods should return arrays, even if they only contain one object. The default Perform code adds an array around the cleaned webhook payload, so if your webhook already provides an array, you can remove this additional array and simply `return bundle.cleanedRequest;`.

Reviewer: I'd rephrase:

```suggestion
Both the Perform and Perform List methods should return an array, even if they only contain one object. The default Perform code adds an array around the cleaned webhook payload, as in `return [bundle.cleanedRequest];`, so if your webhook already provides an array, you can simply `return bundle.cleanedRequest;`.
```

Author: You're right that arrays isn't clear here, but there are numeric agreement issues with `an array` with the rest of the sentence, and the default code is mentioned in the previous paragraph, so I may make a slightly different adjustment on this.

If, for architectural reasons, your webhook will receive some data that shouldn't trigger the Zap, your code can return an empty array in those cases. If the Perform method returns an empty array, the Zap won't run.

For data sent to Zapier via REST Hook, most requests will be successful and return a 200 status code with some request-tracking data. This indicates that Zapier has accepted the data, but it's still possible for errors to occur within the Zap if the structure of the provided data is unexpected.

When sending data:

* Be mindful of Zapier's [rate limits](https://zapier.com/help/troubleshoot/behavior/rate-limits-and-throttling-in-zapier#step-4).
* If your application receives a 410 response, that webhook subscription is no longer active, and you should stop sending data to it.
* If your application receives repeated 4xx or 5xx failures from Zapier outside those error types, this can be handled at your discretion. You may choose to try again later, or to stop sending data and deactivate the hook.

> To signal that your app has deactivated the hook for any reason, you can optionally send a "reverse unsubscribe" call to Zapier. This can allow users to manage their subscriptions from inside your app, or permit you to clean up after a user deletes their account or revokes credentials. The request should be of the form `DELETE <target_url>`, where `target_url` is the unique target URL that was provided when the subscription was created. This will pause the Zap within Zapier.

Reviewer: TIL! I wonder why it is called "reverse unsubscribe". Isn't "reverse unsubscribe" "subscribe"?

Author: I think it's because it's an unsubscribe from the opposite direction 😄 But it is a bit funny in terms of terminology.
